rust/lance/src/dataset.rs

```diff
@@ -434,6 +439,30 @@ impl Dataset {
             });
         }
 
+        // If indices were also the last block, we can take the opportunity to
+        // decode them now and cache them.
+        if let Some(index_offset) = manifest.index_section {
+            if manifest_size - index_offset <= last_block.len() {
+                let offset_in_block = last_block.len() - (manifest_size - index_offset);
+                let message_len =
+                    LittleEndian::read_u32(&last_block[offset_in_block..offset_in_block + 4])
+                        as usize;
+                let message_data =
+                    &last_block[offset_in_block + 4..offset_in_block + 4 + message_len];
+                let section = lance_table::format::pb::IndexSection::decode(message_data)?;
+                let indices = section
+                    .indices
+                    .into_iter()
+                    .map(Index::try_from)
+                    .collect::<Result<Vec<_>>>()?;
+                session.index_cache.insert_metadata(
+                    base_path.as_ref(),
+                    manifest_location.version,
+                    Arc::new(indices),
+                );
+            }
+        }
```

Comment:
Nice.  I've been wondering lately if we should just bite the bullet and make the high level index metadata a part of the manifest proper instead of a secondary message that requires independent loading, just for simplicity of understanding and not performance.

Reply:
Yeah it would be interesting to see if it's worth splitting that out in the first place.